### PR TITLE
8324584: Optimize Symbol and char* handling in ClassLoader

### PR DESCRIPTION
**Pull request by matias9927**
The call to `ClassLoader::load_class()` currently takes a Symbol* for the name and later extracts the string only to change it back into a Symbol* inside `ClassLoader::search_module_entries()` while also recalculating the package entry. This patch simply passes the string directly along with the package entry from the caller. Verified with tier1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324584](https://bugs.openjdk.org/browse/JDK-8324584): Optimize Symbol and char* handling in ClassLoader (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17761/head:pull/17761` \
`$ git checkout pull/17761`

Update a local copy of the PR: \
`$ git checkout pull/17761` \
`$ git pull https://git.openjdk.org/jdk.git pull/17761/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17761`

View PR using the GUI difftool: \
`$ git pr show -t 17761`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17761.diff">https://git.openjdk.org/jdk/pull/17761.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17761#issuecomment-1933040348)

### CODE REVIEW

File: src/hotspot/share/classfile/classLoader.cpp [hunk 1]
```diff
@@ -1031,14 +1031,12 @@ ClassPathEntry* find_first_module_cpe(ModuleEntry* mod_entry,
 // Search either the patch-module or exploded build entries for class.
 ClassFileStream* ClassLoader::search_module_entries(JavaThread* current,
                                                     const GrowableArray<ModuleClassPathList*>* const module_list,
-                                                    const char* const class_name,
+                                                    Symbol* name,
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoader.cpp:1034]: AFAICS `name` is never actually used. The class name was only used to get the package entry, which we already have.

The comment on L1031 also needs updating as no class is involved.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Thanks for taking on this cleanup.

I think there is a further simplification here.

File: src/hotspot/share/classfile/classLoader.cpp (hunk 1 above)
Comment (matias9927) [reply src/hotspot/share/classfile/classLoader.cpp]: Good catch!

**Review by coleenp — APPROVED** [review]
This looks like a good cleanup.  Put noreg-hard on the bug because it's not possible to write a test for this.
Update the copyright years if they haven't been already.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Just a few requests to explain the pkg_entry role. Thanks

File: src/hotspot/share/classfile/classLoader.hpp [hunk 2]
```diff
@@ -295,11 +295,11 @@ class ClassLoader: AllStatic {
   // Attempt load of individual class from either the patched or exploded modules build lists
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/classfile/classLoader.hpp:295]: Actually this comment is wrong for this method.

**Review by coleenp — APPROVED** [review]
Still looks good!

**Review by dholmes-ora — COMMENTED** [review]
I hate to belabour this, but in trying to understand what the `pkg_entry` means it seems to me that it is only needed to extract the `ModuleEntry` and we have that information available at the top-level too. So we could pass that through instead. Thoughts?

The problem I'm having with this is how you would document/explain the API, because we are basically passing through "tramp data" for convenience as we evaluated something higher up the call chain that is needed again lower-down, so we pass it through rather than re-evaluating/re-calculating. But that then means there are very strict constraints between the different parameters, which are not captured/documented anywhere.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the offline discussion @matias9927. I see now that this aligns with the existing `load_shared_class` API (which also doesn't document its arguments and their constraints very well). There is room for some further improvement here but that would be a different RFE.

Thanks.

**Comment by matias9927** [comment]
Thanks for the reviews @coleenp and @dholmes-ora!
/integrate

**Pull request closed without merging**